modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxLocal.java

```diff
@@ -3744,10 +3809,121 @@ private void readyNearLock(IgniteTxEntry txEntry,
         return true;
     }
 
+    /**
+     * Returns current amount of time that transaction has spent on system activities (acquiring locks, commiting,
+     * rolling back, etc.)
+     *
+     * @return Amount of time in milliseconds.
+     */
+    public long systemTimeCurrent() {
+        long systemTime0 = systemTime.get();
+
+        long systemStartTime0 = systemStartTime.get();
+
+        long t = systemStartTime0 == 0 ? 0 : (System.nanoTime() - systemStartTime0);
+
+        return U.nanosToMillis(systemTime0 + t);
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean state(TransactionState state) {
+        boolean res = super.state(state);
+
+        if (state == COMMITTED || state == ROLLED_BACK) {
+            leaveSystemSection();
+
+            //if commitOrRollbackTime != 0 it means that we already have written metrics and dumped it in log at least once
+            if (!commitOrRollbackTime.compareAndSet(0, System.nanoTime() - commitOrRollbackStartTime.get()))
+                return res;
+
+            long systemTimeMillis = U.nanosToMillis(this.systemTime.get());
+            long totalTimeMillis = System.currentTimeMillis() - startTime();
+            long userTimeMillis = totalTimeMillis - systemTimeMillis;
+
+            writeTxMetrics(systemTimeMillis, userTimeMillis);
+
+            long transactionTimeDumpThreshold = cctx.tm().longTransactionTimeDumpThreshold();
+
+            double transactionTimeDumpSamplesCoefficient = cctx.tm().transactionTimeDumpSamplesCoefficient();
+
+            boolean isLong = transactionTimeDumpThreshold > 0 && totalTimeMillis > transactionTimeDumpThreshold;
+
+            boolean randomlyChosen = transactionTimeDumpSamplesCoefficient >= 0.0
+                && RANDOM.nextDouble() <= transactionTimeDumpSamplesCoefficient
+                && transactionDumpHitRateCounter != null
```

Comment:
Why `transactionDumpHitRateCounter` could be null? You should add `@Nullable` to costructor args or remove null check from condition.

Reply:
Because there is a default constructor and also mock child class that uses multiple null arguments for constructor.